graph/src/components/store.rs

```diff
@@ -245,7 +276,111 @@ where
             })
         });
 
-        StoreEventStream { source }
+        StoreEventStream::new(Box::new(source))
+    }
+
+    /// Reduce the frequency with which events are generated while a
+    /// subgraph deployment is syncing. While the given `deployment` is not
+    /// synced yet, events from `source` are reported at most every
+    /// `interval`. At the same time, no event is held for longer than
+    /// `interval`. The `StoreEvents` that arrive during an interval appear
+    /// on the returned stream as a single `StoreEvent`; the events are
+    /// combined by using the maximum of all sources and the concatenation
+    /// of the changes of the `StoreEvents` received during the interval.
+    pub fn throttle_while_syncing(
+        self,
+        store: Arc<impl Store>,
+        deployment: SubgraphDeploymentId,
+        interval: Duration,
+    ) -> StoreEventStreamBox {
+        // We refresh the synced flag every SYNC_REFRESH_FREQ*interval to
+        // avoid hitting the database too often to see if the subgraph has
+        // been synced in the meantime. The only downside of this approach is
+        // that we might continue throttling subscription updates for a little
+        // bit longer than we really should
+        static SYNC_REFRESH_FREQ: u32 = 4;
+
+        let mut synced = false;
+        let synced_check_interval = interval.checked_mul(SYNC_REFRESH_FREQ).unwrap();
```

Comment:
Do you mean the `unwrap` should go away? It's highly unlikely that we'd get to an overflow here since that would require setting `interval` to something insanely large.

Reply:
Nothing to change, it's great that you're being mindful of this.